src/middlewared/middlewared/plugins/vm.py

```diff
@@ -564,6 +824,94 @@ def identify_hypervisor(self):
+    async def fetch_image(self, job, vmOS, force=False):
+        """Download a pre-built image for bhyve"""
+        vm_os = CONTAINER_IMAGES.get(vmOS)
+        url = vm_os['URL']
+
+        self.logger.debug("==> IMAGE: {0}".format(vm_os))
+
+        sharefs = await self.middleware.call('vm.get_sharefs')
+        vm_os_file = vm_os['GZIPFILE']
+        iso_path = sharefs + '/iso_files/'
+        file_path = iso_path + vm_os_file
+
+        if os.path.exists(file_path) is False and force is False:
+            logger.debug("===> Downloading: %s" % (url))
+            await self.middleware.threaded(lambda: urlretrieve(url, file_path,
+                lambda nb, bs, fs, job=job: self.fetch_hookreport(nb, bs, fs, job, file_path)))
+
+    @accepts()
+    async def list_images(self):
+        return CONTAINER_IMAGES
+
+    @accepts(Int('job_id'))
+    async def get_download_status(self, job_id):
+        """ Returns the status of the job, if job does not exists it returns False."""
+        job_pool = await self.middleware.call('core.get_jobs', [('method', '=', 'vm.fetch_image')])
+        for job in job_pool:
+            if job['id'] == job_id:
+                return job
+        return False
+
+    @accepts(Str('vmOS'))
+    async def image_path(self, vmOS):
+        """Return the prebuilt image path or false in case it is not supported."""
+        vm_os = CONTAINER_IMAGES.get(vmOS, None)
+        if vm_os:
+            image_file = vm_os['GZIPFILE']
+            sharefs = await self.middleware.call('vm.get_sharefs')
+            file_path = sharefs + '/iso_files/' + image_file
+            if os.path.exists(file_path):
+                if self.vmutils.check_sha256(file_path, vmOS):
+                    self.logger.debug("===> Checksum OK: {}".format(file_path))
+                    return file_path
+                else:
+                    self.logger.debug("===> Checksum NOK: {}".format(file_path))
+                    return False
+            else:
+                return False
+        else:
+            return False
+
+    def decompress_hookreport(self, dst_file, job):
+        totalsize = 4756340736  # XXX: It will be parsed from a sha256 file.
+        fd = os.open(dst_file, os.O_RDONLY)
+        try:
+            size = os.lseek(fd, 0, os.SEEK_END)
+        finally:
+            os.close(fd)
+
+        percent = (size / totalsize) * 100
+        job.set_progress(int(percent), 'Decompress', {'decompressed': size, 'total': totalsize})
+
+    @accepts(Str('src'), Str('dst'))
+    def decompress_gzip(self, src, dst):
+        if os.path.exists(dst):
+            self.logger.error("===> DST: {0} exist, we stop here.".format(dst))
+            return False
+
+        if self.vmutils.is_gzip(src) is True:
+            self.logger.debug("===> SRC: {0} DST: {1}".format(src, dst))
+            src_file = gzip.open(src, 'rb')
+            dst_file = open(dst, 'wb')
+            dst_file.write(src_file.read())
+            src_file.close()
```

Comment:
Using `with` operator for resources such as open files will result in less code, will be more readable and will provide proper cleanup in case of exception